Import & update the collection 3 product definitions

jeremyh opened the pull request:
Imported from @Kirill888 's originals in [dea-proto](https://github.com/opendatacube/dea-proto) with git history intact.

Made changes based on the feedback register:

- Added simple aliases for red/green/blue etc for nbart.
- Expanded the product descriptions (A little. Maybe people will want more?)
- There's now multiple metadata types (and a minimal default 'eo3') so that products wont have a large number of confusing null fields. (original feedback was for Level1, but it will affect all non-ARD products)
- Added aliases without `oa_` prefixes
- Fixed invalid nodata values (uint8 -> -999)
- Added the new naming conventions for products `*.odc-product.yaml`

Misc changes
- Removed unnecessary indexes (for platform/instrument/etc), which don't differ within a product.
- Ran a ~ruamel whitespace~ headless pycharm autoformat on the whole history.

Still outstanding:
- ~nodata value review~ (✓done)
- a second/fmask cloud field (needs to be pushed upstream to the datasets first if we're doing it)

Kirill888 committed:
Draft metadata definition for DEA collection 3

Kirill888 committed:
Consistent order of lonlat in EO3 spec

Kirill888 committed:
Adding ARD product definitions for LS5/7/8

Kirill888 committed:
Change region_code location (odc:reference_code -> odc:region_code)

Kirill888 committed:
Add product definitions for L1 datasets

Kirill888 committed:
Expose most properties in EO3 metadata for now

jeremyh committed:
LS8 L1 instrument correction

To match the new datasets.

jeremyh committed:
Expand eo3 product descriptions

As requested in the feedback.

jeremyh committed:
Separate metadata type of Level 1

As requested in feedback: there's a huge
number of null fields as most only exist in ARD

jeremyh committed:
Make the default eo3 format minimal

Most product will not duplicate all the gqa/etc
fields that ARD uses.

jeremyh committed:
Put eo3 products in a common directory

jeremyh committed:
Add eo3 index test, fix l1 reference

jeremyh committed:
Re-introduce simple aliases for nbart bands

As requested by @robbibt in feedback

Kirill888 commented on digitalearthau/config/eo3/products/ard_ls5.yaml:
```diff
@@ -0,0 +1,177 @@
+    units: '1'
+-   name: oa_combined_terrain_shadow
+    dtype: uint8
+    nodata: -999
```
this one needs to be fixed as raised by @simonaoliver in other products too

jeremyh replied:
:+1:

Kirill888 commented on digitalearthau/config/eo3/products/ard_ls7.yaml:
```diff
@@ -0,0 +1,190 @@
+    units: '1'
+-   name: oa_combined_terrain_shadow
+    dtype: uint8
+    nodata: -999
```
nodata value outside of dtype range

jeremyh replied:
:+1:

Kirill888 commented on digitalearthau/config/eo3/products/ard_ls8.yaml:
```diff
@@ -0,0 +1,203 @@
+    units: '1'
+-   name: oa_combined_terrain_shadow
+    dtype: uint8
+    nodata: -999
```
and again

jeremyh replied:
:+1:

Kirill888 commented on digitalearthau/config/eo3/metadata_eo3.yaml:
```diff
@@ -0,0 +1,85 @@
+name: eo3
+description: Default EO3 with no custom fields
+dataset:
+    id: [id]  # needs to be fixed
```
I know I put this comment in myself `# needs to be fixed` but it can be misinterpreted as "needs fixing" while it really means "has to be this value and not any other"

jeremyh replied:
Updated the comment

jeremyh committed:
Compact yaml format must be quoted

According to yamllint. Cleanup warnings.

jeremyh committed:
Combined-terrain-shadow should have nodata 255

It was outside the range of uint8.

From the feedback register.

jeremyh committed:
Naming conventions: odc-product.yaml, odc-type.yaml

As discussed on Slack to match *.odc-dataset.yaml.

jeremyh committed:
Tighten glob to match naming conventions

jeremyh committed:
OA measurements should have aliases without OA

Requested in feedback, approved in Slack.

jeremyh committed:
Add aliases for swir1/swir2 to match previous collection

Requested by @robbibt on Slack. Approved by S.Oliver.

(the actual names cannot be changed easily as they must match the datasets upstream. And underscores are seen
as more consistent with other names.)

jeremyh committed:
Re-add comments to non-modifiable md-type fields

jeremyh commented:
Products now validate against the sample data:

```
$ eo3-validate --thorough /g/data/v10/agdc/jez/digitalearthau/digitalearthau/config/eo3/products/*.yaml `fd odc-metadata.yaml`
✓ ard_ls5
✓ ard_ls7
✓ ard_ls8
✓ l1_ls5
✓ l1_ls7
✓ l1_ls8
✓ ga_ls5t_ard_3-0-0_092084_2009-12-17_final.odc-metadata
	I different_nodata oa_combined_terrain_shadow nodata: product 255 != dataset None
✓ ga_ls7e_ard_3-0-0_092084_2016-08-23_final.odc-metadata
	I different_nodata oa_combined_terrain_shadow nodata: product 255 != dataset None
✓ ga_ls8c_ard_3-0-0_092084_2016-06-28_final.odc-metadata
	I different_nodata oa_combined_terrain_shadow nodata: product 255 != dataset None

valid: 3 infos
```

The `combined_terrain_shadow` dataset nodata is None according to @sixy6e as it never has any nodata values, and 255 will be used in the product as the ODC filll value. 

(Because of the fill-value/different usage of these fields, I've changed the validator to allow nodata values to differ, only printing an info)

jeremyh commented:
(in keeping the Git history of this branch clean I've made this issue history a mess :slightly_smiling_face: )

Kirill888 reviewed (approved):
:+1:

jeremyh committed:
Add 'digitalearthau/config/eo3/' from commit 'a62c2e6554cd7ad9f23ba15eae5483910cd421e6'

git-subtree-dir: digitalearthau/config/eo3
git-subtree-mainline: 1860ba1b1a4dd9141165576d65b42d04711131e3
git-subtree-split: a62c2e6554cd7ad9f23ba15eae5483910cd421e6

jeremyh committed:
Rename tmad product to match conventions
